src/accounts.rs

```diff
@@ -175,18 +178,67 @@ impl AccountsDB {
         }
     }
 
+    fn load_executable_accounts(&self, mut program_id: Pubkey) -> Result<Vec<(Pubkey, Account)>> {
+        let mut accounts = Vec::new();
+        let mut depth = 0;
+        loop {
+            if solana_native_loader::check_id(&program_id) {
+                // at the root of the chain, ready to dispatch
+                break;
+            }
+
+            if depth >= 5 {
+                return Err(BankError::CallChainTooDeep);
+            }
+            depth += 1;
+
+            let program = match self.load(&program_id) {
+                Some(program) => program,
+                None => return Err(BankError::AccountNotFound),
+            };
+            if !program.executable || program.loader == Pubkey::default() {
+                return Err(BankError::AccountNotFound);
+            }
+
+            // add loader to chain
+            accounts.insert(0, (program_id, program.clone()));
+
+            program_id = program.loader;
+        }
+        Ok(accounts)
+    }
+
+    /// For each program_id in the transaction, load its loaders.
+    fn load_loaders(&self, tx: &Transaction) -> Result<Vec<Vec<(Pubkey, Account)>>> {
+        tx.instructions
+            .iter()
+            .map(|ix| {
+                let program_id = tx.program_ids[ix.program_ids_index as usize];
+                self.load_executable_accounts(program_id)
+            })
+            .collect()
+    }
+
     fn load_accounts(
         &self,
         txs: &[Transaction],
         last_ids: &mut StatusDeque<Result<()>>,
-        results: Vec<Result<()>>,
+        lock_results: Vec<Result<()>>,
         max_age: usize,
         error_counters: &mut ErrorCounters,
-    ) -> Vec<(Result<Vec<Account>>)> {
+    ) -> Vec<Result<(InstructionAccounts, InstructionLoaders)>> {
         txs.iter()
-            .zip(results.into_iter())
+            .zip(lock_results.into_iter())
             .map(|etx| match etx {
```

Comment:
Can you either break this out into a separate function or find a way to make use of the `?` operator to get rid of all the pattern matching?

Reply:
@garious, result matching gotten rid of :-)